src/java.base/unix/classes/sun/nio/fs/UnixFileAttributeViews.java

```diff
@@ -76,13 +76,16 @@ public void setTimes(FileTime lastModifiedTime,
             boolean useFutimes = false;
             boolean useFutimens = false;
             boolean useLutimes = false;
+            boolean useUtimensat = false;
             int fd = -1;
             try {
-                if (!followLinks) {
-                    useLutimes = lutimesSupported() &&
-                        UnixFileAttributes.get(file, false).isSymbolicLink();
+                if (!followLinks &&
+                    UnixFileAttributes.get(file, false).isSymbolicLink()) {
+                    useUtimensat = utimensatSupported();
+                    if (!useUtimensat)
+                        useLutimes = lutimesSupported();
                 }
-                if (!useLutimes) {
+                if (!useUtimensat && !useLutimes) {
```

Comment:
Indeed UnixFileAttributes.get calls lstat. Calling symlink.openForAttributeAccess(false) however yields
```
java.nio.file.FileSystemException: symlink: No such file or directory or unable to access attributes of symbolic link
```
According to [symlink(7)](https://www.man7.org/linux/man-pages/man7/symlink.7.html):
```
The last access and last modification timestamps of a symbolic link can be changed using utimensat(2) or lutimes(3).
```
It looks like futimens will not work with a symlink's fd so that utimensat is necessary here to obtain nsec precision. I don't see a way to close the window between lstat and utimensat, which anyway was already there for the lutimes case.

Reply:
I don't think that the logic in 08f5b7d is quite correct yet. It might be possible if followLinks is false and the file is not a symlink, that lutimes would be used even though futimens is supported, thus decreasing the timestamp resolution.